src/schemas/m2cms.ts

```diff
@@ -0,0 +1,152 @@
+import process from 'process';
+import fs from 'node:fs';
+import path from 'node:path';
+import Typescript, {
+    type Node,
+    type InterfaceDeclaration,
+    type PropertySignature,
+    type Identifier
+} from 'typescript';
+import YAML from 'yaml';
+
+import Debug from 'debug';
+const DEBUG = Debug('m2cms')
+
+const KIND = (e: Node | undefined ) => e && Typescript.SyntaxKind[e.kind]
+
+interface Field {
+    label: string;
+    name: string;
+    widget: string;
+    fields?: Field[];
+}
+
+interface Collection {
+    name: string;
+    label: string;
+    folder: string;
+    widget: string;
+    create: boolean;
+    slug: string;
+    fields: Field[];
+}
+
+interface Config {
+    local_backend?: boolean;
+    backend?: {
+        name: string;
+        repo: string;
+        branch?: string;
+    }
+    publish_mode?: string;
+    media_folder: string;
+    collections: Collection[];
+}
+
+
+function parsePropertySignature(ps: PropertySignature, ancestors: string[] = []) : Field | null {
+    const name = (ps.name as Identifier).escapedText.toString();
+    const config = {
+        name,
+        label: [...ancestors, name].join(' '),
+        required: !!!ps.questionToken
+    }
+
+    switch(ps && ps.type && ps.type.kind) {
+        case Typescript.SyntaxKind.StringKeyword:
+            return {
+                ...config,
+                widget: "string"
+            }
+        case Typescript.SyntaxKind.NumberKeyword:
+            return {
+                ...config,
+                widget: "number"
+            }
+        case Typescript.SyntaxKind.ObjectKeyword:
+            return {
+                ...config,
+                widget: "object"
+            }
+        case Typescript.SyntaxKind.TupleType:
+            return {
+                ...config,
+                widget: "list"
+            }
+        case Typescript.SyntaxKind.TypeLiteral:
+            return parseNode(ps, [...ancestors, name])
+        case Typescript.SyntaxKind.TypeReference:
+            return collections[name]
+        default:
+            DEBUG('unhandled type', KIND(ps.type), ps, collections)
+            throw "Type Error"
+    }
+    return null;
+}
+
+function parseNode(node : Node, ancestors: string[] = []): Collection {
+
+    const fields : Field[] = []
+    node.forEachChild((child: Node) => {
+        switch (child.kind) {
+            case Typescript.SyntaxKind.InterfaceDeclaration:
```

Comment:
> Speaking of casts, parseNode has two of them that can be avoided! In this case, if statements (rather than a switch) can provide better type safety with type guards such as isPropertySignature which come with Typescript.

typescript being typescript.